src/ffi.rs

```diff
@@ -223,42 +223,47 @@ impl ForeignFunctionTable {
         arg: &mut Value,
         structs_table: &mut HashMap<String, StructImpl>,
     ) -> Result<(Box<dyn Any>, usize, usize), FFIError> {
-        unsafe {
-            match arg {
-                Value::Struct(ref name, ref mut struct_args) => {
-                    if let Some(ref mut struct_type) = structs_table.clone().get_mut(name) {
-                        let layout = Layout::from_size_align(
-                            struct_type.ffi_type.size,
-                            struct_type.ffi_type.alignment.into(),
-                        )
-                        .unwrap();
-                        let align = struct_type.ffi_type.alignment as usize;
-                        let size = struct_type.ffi_type.size;
-                        let ptr = alloc(layout) as *mut c_void;
-                        let mut field_ptr = ptr;
-
-                        #[allow(clippy::needless_range_loop)]
-                        for i in 0..(struct_type.fields.len() - 1) {
-                            macro_rules! try_write_int {
-                                ($type:ty) => {{
-                                    field_ptr = field_ptr
-                                        .add(field_ptr.align_offset(std::mem::align_of::<$type>()));
-                                    let n: $type = <$type>::try_from(struct_args[i].as_int()?)
-                                        .map_err(|_| FFIError::ValueDontFit)?;
-                                    std::ptr::write(field_ptr as *mut $type, n);
-                                    field_ptr = field_ptr.add(std::mem::size_of::<$type>());
-                                }};
-                            }
+        match arg {
+            Value::Struct(ref name, ref mut struct_args) => {
+                if let Some(ref mut struct_type) = structs_table.clone().get_mut(name) {
+                    let layout = Layout::from_size_align(
+                        struct_type.ffi_type.size,
+                        struct_type.ffi_type.alignment.into(),
+                    )
+                    .unwrap();
+                    let align = struct_type.ffi_type.alignment as usize;
+                    let size = struct_type.ffi_type.size;
+                    let ptr = unsafe { alloc::alloc(layout) as *mut c_void };
 
-                            macro_rules! write {
-                                ($type:ty, $value:expr) => {{
-                                    let data: $type = $value;
-                                    std::ptr::write(field_ptr as *mut $type, data);
-                                    field_ptr = field_ptr.add(align);
-                                }};
-                            }
+                    if ptr.is_null() {
+                        panic!("allocation failed")
+                    }
```

Comment:
You might want to hide whitespace changes when reviewing this diff as shrinking the function spanning unsafe block into three minimal sized unsafe blocks introduced a lot of whitespace change.